pubsublite/streams.go

```diff
@@ -0,0 +1,333 @@
+//
+// A retryableStream cycles between the following goroutines:
+//   Start() --> reconnect() <--> listen()
+// terminate() can be called at any time, either by the client to force stream
+// closure, or as a result of an unretryable error.
+type retryableStream struct {
+	// Immutable after creation.
+	ctx          context.Context
+	handler      streamHandler
+	responseType reflect.Type
+	timeout      time.Duration
+
+	// Guards access to fields below.
+	mu sync.Mutex
+
+	stream       grpc.ClientStream
+	cancelStream context.CancelFunc
+	status       streamStatus
+	finalErr     error
+}
+
+// newRetryableStream creates a new retryable stream wrapper. `timeout` is the
+// maximum duration for reconnection. `responseType` is the type of the response
+// proto received on the stream.
+func newRetryableStream(ctx context.Context, handler streamHandler, timeout time.Duration, responseType reflect.Type) *retryableStream {
+	return &retryableStream{
+		ctx:          ctx,
+		handler:      handler,
+		responseType: responseType,
+		timeout:      timeout,
+	}
+}
+
+// Start establishes a stream connection and returns the result to the given
+// channel.
+func (rs *retryableStream) Start(result chan error) {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+
+	if rs.status != streamUninitialized {
+		result <- status.Error(codes.Internal, "pubsublite: stream has already been started")
+		return
+	}
+	go rs.reconnect(&result)
+}
+
+// Stop gracefully closes the stream without error.
+func (rs *retryableStream) Stop() {
+	rs.terminate(nil)
+}
+
+// Send attempts to send the request to the underlying stream and returns true
+// if successfully sent. Returns false if an error occurred or a reconnection is
+// in progress.
+func (rs *retryableStream) Send(request interface{}) (sent bool) {
+	rs.mu.Lock()
+
+	if rs.stream != nil {
+		err := rs.stream.SendMsg(request)
+		// Note that if SendMsg returns an error, the stream is aborted.
+		switch {
+		case err == nil:
+			sent = true
+		case err == io.EOF:
+			// If SendMsg returns io.EOF, RecvMsg will return the status of the
+			// stream. Nothing to do here.
+			break
+		case isRetryableSendError(err):
+			go rs.reconnect(nil)
+		default:
+			rs.mu.Unlock() // terminate acquires the mutex.
+			rs.terminate(err)
+			return
+		}
+	}
+
+	rs.mu.Unlock()
+	return
+}
+
+// Status returns the current status of the stream.
+func (rs *retryableStream) Status() streamStatus {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	return rs.status
+}
+
+// Error returns the error that caused the stream to terminate. Can be nil if it
+// was initiated by Stop().
+func (rs *retryableStream) Error() error {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	return rs.finalErr
+}
+
+func (rs *retryableStream) currentStream() grpc.ClientStream {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	return rs.stream
+}
+
+// clearStream must be called with the retryableStream.mu locked.
+func (rs *retryableStream) clearStream() {
+	if rs.cancelStream != nil {
```

Comment:
Does this need to be guarded by the mutex?

Reply:
Yes it does, I added the comment above about the precondition, but wasn't sure how to/whether to enforce it.
clearStream() needs to be atomic with other operations and used by a couple of methods.